### PR TITLE
Adds ruff linter, fixes unused variables errors

### LINKED ISSUE #8665: Start to use Pylint
Issues as https://github.com/DefectDojo/django-DefectDojo/issues/8636#issuecomment-1719081247 (fixed in https://github.com/DefectDojo/django-DefectDojo/pull/8663) would not happen if we would have a track of unused variables.

I suggest using Pylint to keep the code clean. Flake8 does a lot but it is not enough.

### PR DESCRIPTION
**Pull request by dwalleck**
**Description**

closes #8665 

This change adds the ruff linter, which is a Rust-driven Python tool for timely linting. It covers the same ruleset as flake8 and other Python linting tools. In particular, this change adds a rule to disallow unused variables. Once we can guarantee that ruff is covering everything that flake8 covers, we should be able to remove flake8 from gating.

**Test results**

The ruff tool reports no errors with the specified ruleset.

### CODE REVIEW

**Comment by mtesauro** [comment]
@dwalleck Thanks for this contribution.

This looks very promising.  Please get those couple of failing tests to pass so we can merge this. :+1:

**Comment by kiblik** [comment]
Wow, nice. I suppose it identified many places which will optimise evaluation.

Originally I was scared that DD is not validating passwords on change but I found out it is done in the `form.is_valid()`

I suppose this should not be simply merged because in some cases it will need the removal of the whole line (not only `var =`) and in some other cases usage of data should be added (not only the removal of their processing).

File: dojo/cred/views.py
Comment (Maffooch) [inline dojo/cred/views.py:332]: Agreed with Kiblik ^ here is an example of still doing the unused processing

Comment (dwalleck) [reply dojo/cred/views.py:332]: Good catch. I'll go back over these changes a bit more carefully

File: dojo/cred/views.py
Comment (Maffooch) [inline dojo/cred/views.py:268]: This function could be removed

File: dojo/tools/outpost24/parser.py
Comment (Maffooch) [inline dojo/tools/outpost24/parser.py:62]: This one could be worth keeping to improve the parser

**Review by Maffooch — CHANGES_REQUESTED** [review]
Did a quick sweep

File: dojo/group/utils.py
Comment (kiblik) [reply dojo/group/utils.py:18]:
```suggestion
        Group.objects.get(name=auth_group_name)
```

File: dojo/jira_link/views.py
Comment (kiblik) [inline dojo/jira_link/views.py:329]:
```suggestion
            jira_helper.get_jira_connection_raw(jira_server, jira_username, jira_password)
```

File: dojo/management/commands/migrate_surveys.py
Comment (kiblik) [inline dojo/management/commands/migrate_surveys.py:42]:
```suggestion
            TextQuestion.objects.create(optional=False, order=1, text='What is love?')
```

File: dojo/metrics/views.py
Comment (kiblik) [reply dojo/metrics/views.py:393]:
```suggestion
```

File: dojo/forms.py
Comment (kiblik) [reply dojo/forms.py:3014]:
```suggestion
```

File: dojo/models.py
Comment (kiblik) [reply dojo/models.py:2824]:
```suggestion
            self.github_issue
```

File: dojo/forms.py
Comment (kiblik) [inline dojo/forms.py:663]:
```suggestion
        kwargs.pop('finding')
```

File: dojo/jira_link/helper.py
Comment (kiblik) [inline dojo/jira_link/helper.py:788]:
```suggestion
        jira.issue(new_issue.id)
```

File: dojo/product/views.py
Comment (kiblik) [inline dojo/product/views.py:67]:
```suggestion
            get_object_or_404(Product_Type, id=p[0])
```

File: dojo/product/views.py
Comment (kiblik) [inline dojo/product/views.py:1004]:
```suggestion
        jira_helper.get_jira_project(product)
```

File: dojo/product/views.py
Comment (kiblik) [inline dojo/product/views.py:1063]:
```suggestion
            jira_helper.get_jira_project(product)
```

File: dojo/reports/widgets.py
Comment (kiblik) [inline dojo/reports/widgets.py:31]:
```suggestion
            json.loads(jdata)
```

File: dojo/user/views.py
Comment (kiblik) [inline dojo/user/views.py:184]:
```suggestion
        request.POST.getlist('alert_select')
```

**Review by kiblik — CHANGES_REQUESTED** [review]
Some feedback from my review

**Comment by kiblik** [comment]
Is it possible to approve the workflow here, please? It would be good to know, what kind of damage it creates in tests and unittests

File: .github/workflows/ruff.yml
Comment (dwalleck) [reply .github/workflows/ruff.yml:37]: Those are both great ideas! However, for the sake of not making this PR even larger, I would prefer to add the linting of those directories in a separate PR that quickly follows this one.

File: dojo/api_v2/views.py
Comment (dwalleck) [reply dojo/api_v2/views.py:600]: Thanks! I will remove it

File: dojo/cred/views.py
Comment (dwalleck) [reply dojo/cred/views.py:268]: Thanks. I'll remove it

File: requirements-dev.txt
Comment (dwalleck) [reply requirements-dev.txt:1]: It's very common to have a requirements-dev/test/etc for tools used only in developing an app. Right now there is only one dependency, but there will be many more.

File: .github/workflows/ruff.yml
Comment (kiblik) [reply .github/workflows/ruff.yml:37]: Fully understand. No problem. It is a good idea.

File: requirements-dev.txt
Comment (kiblik) [reply requirements-dev.txt:1]: Ok, no problem

**Review by mtesauro — APPROVED** [review]
Approved.

**Comment by mtesauro** [comment]
@dwalleck I approved assuming you've resolved the pending reviews - the people who put those in should have a look and close those out soon.

@kiblik @Maffooch

File: dojo/group/utils.py
Comment (Maffooch) [inline dojo/group/utils.py:18]: This can be removed now

**Review by Maffooch — CHANGES_REQUESTED** [review]
Commenting on the whole review instead of each instance. For instances of removing an unused variable, please also check out what the function is doing to determine if it can also be removed. In one instance I found, there are things like this:
Before:
```
engagements = Engagement.objects.filter(
             product__prod_type=product_type, test__finding__in=findings.qs
         ).distinct()
```
After:
```
Engagement.objects.filter(
             product__prod_type=product_type, test__finding__in=findings.qs
         ).distinct()
```
Though this filter call does not actually result in a DB query due to lazy eval by django, this is now just dead code that serves no purpose

**Review by kiblik — CHANGES_REQUESTED** [review]
Sorry guys, I prepared my review but forgot to submit it.

**Pull request closed without merging**
